truffle/src/main/java/org/jruby/truffle/nodes/objects/AllocateObjectNode.java

```diff
@@ -52,17 +61,62 @@ public DynamicObject allocateCached(
     }
 
     @CompilerDirectives.TruffleBoundary
-    @Specialization(contains = "allocateCached", guards = "!isSingleton(classToAllocate)")
+    @Specialization(
+            contains = "allocateCached",
+            guards = "!isSingleton(classToAllocate)",
+            assumptions = "getTracingAssumption()")
     public DynamicObject allocateUncached(DynamicObject classToAllocate, Object[] values) {
         return getInstanceFactory(classToAllocate).newInstance(values);
     }
 
+    @CompilerDirectives.TruffleBoundary
+    @Specialization(guards = {"!isSingleton(classToAllocate)", "isTracing()"})
+    public DynamicObject allocateTracing(DynamicObject classToAllocate, Object[] values) {
+        final DynamicObject object = getInstanceFactory(classToAllocate).newInstance(values);
+
+        final Node caller = RubyCallStack.getTopMostUserCallNode();
+        final SourceSection callerSource = caller.getEncapsulatingSourceSection();
+
+        final String callerMethod;
+
+        if (caller.getRootNode() instanceof RubyRootNode) {
+            callerMethod = ((RubyRootNode) caller.getRootNode()).getSharedMethodInfo().getName();
+        } else {
+            callerMethod = "(unknown)";
+        }
+
+        getContext().getObjectSpaceManager().traceAllocation(
+                object,
+                string(Layouts.CLASS.getFields(classToAllocate).getName()),
+                string(callerMethod),
+                string(callerSource.getSource().getShortName()),
+                callerSource.getStartLine());
+
+        return object;
+    }
```

Comment:
Nice!
After activating and desactivating allocation tracing, will the fast path be used again?